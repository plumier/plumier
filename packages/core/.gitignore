!src/generic-controller/filter-grammar.js