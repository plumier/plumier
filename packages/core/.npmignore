test
src
.npmignore
.gitignore
tsconfig.json
package-lock.json
tsconfig.build.json
yarn-error.log
yarn.lock