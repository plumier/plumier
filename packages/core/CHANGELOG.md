# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.1.2](https://github.com/plumier/plumier/compare/v1.1.1...v1.1.2) (2023-01-20)

**Note:** Version bump only for package @plumier/core





## [1.1.1](https://github.com/plumier/plumier/compare/v1.1.0...v1.1.1) (2023-01-03)

**Note:** Version bump only for package @plumier/core





# [1.1.0](https://github.com/plumier/plumier/compare/v1.0.6...v1.1.0) (2022-12-19)


### Features

* Add feature to reflect function and lambda function ([#1029](https://github.com/plumier/plumier/issues/1029)) ([f65ec5b](https://github.com/plumier/plumier/commit/f65ec5baa14af9b34793a2ab74f440233f84c6d7))





## [1.0.6](https://github.com/plumier/plumier/compare/v1.0.5...v1.0.6) (2021-10-04)


### Bug Fixes

* **core:** Fix http status error response for 403 Forbidden and 401 Unauthorized ([#1014](https://github.com/plumier/plumier/issues/1014)) ([b235819](https://github.com/plumier/plumier/commit/b2358195477eaafb0654e92c8605bc1ecf7a8350))





## [1.0.5](https://github.com/plumier/plumier/compare/v1.0.4...v1.0.5) (2021-07-25)

**Note:** Version bump only for package @plumier/core





## [1.0.3](https://github.com/plumier/plumier/compare/v1.0.2...v1.0.3) (2021-06-28)


### Bug Fixes

* Custom validator not being executed on relation property due to conflict with converters ([#977](https://github.com/plumier/plumier/issues/977)) ([af94bc7](https://github.com/plumier/plumier/commit/af94bc7d38c999dfd3c1509b34185e3384eed858))
* Fix duplicate entity policy name on route analysis ([#990](https://github.com/plumier/plumier/issues/990)) ([6a18218](https://github.com/plumier/plumier/commit/6a182186de06555adcb4d7864d177ee8a4df9413))
* Fix entity policy analysis messages on missing entity provider ([#964](https://github.com/plumier/plumier/issues/964)) ([5ac4867](https://github.com/plumier/plumier/commit/5ac48676d2c1e0834c86b2c96642b2dd0ad11cac))
* Fix issue when override Open API Tags on first class entity and first class entity relation ([#988](https://github.com/plumier/plumier/issues/988)) ([3aa2337](https://github.com/plumier/plumier/commit/3aa23377209e8678f7d52e53b3b2196f8d719c2d))
* Unable to cleanly disable authorization (writeonly and readonly) on property ([#976](https://github.com/plumier/plumier/issues/976)) ([bcd544b](https://github.com/plumier/plumier/commit/bcd544b68e2c6a6a9f5c9e33a54f8098b4b06d65))





## [1.0.2](https://github.com/plumier/plumier/compare/v1.0.0...v1.0.2) (2021-06-06)


### Bug Fixes

* Fix array relation should not accessible (read/write) on first class entity ([#956](https://github.com/plumier/plumier/issues/956)) ([58058a5](https://github.com/plumier/plumier/commit/58058a54861447d04cedfd585d60687eb3d4e1d4))
* Fix authorization evaluation issue on undefined value should be skipped ([#945](https://github.com/plumier/plumier/issues/945)) ([7e9acd3](https://github.com/plumier/plumier/commit/7e9acd330032f7f829ef738668768daf4c379566))
* Fix response authorization error when no ID provided on response ([#948](https://github.com/plumier/plumier/issues/948)) ([2b5429e](https://github.com/plumier/plumier/commit/2b5429ef30f9cfb3843fb07c5af271dd3223b14c))
* Fix Swagger urls not visible on route analysis ([#935](https://github.com/plumier/plumier/issues/935)) ([e858197](https://github.com/plumier/plumier/commit/e8581971087ddde0d3642aaa930ac36eafc9bc26))
* Give proper error message when action return type array, but got non array ([#934](https://github.com/plumier/plumier/issues/934)) ([05cd377](https://github.com/plumier/plumier/commit/05cd377823e789e1c18c3902cad69798f196549e))
* Give proper error message when found cross reference entity issue ([#933](https://github.com/plumier/plumier/issues/933)) ([0d09d22](https://github.com/plumier/plumier/commit/0d09d22589b751f54c0c4bd03de9f0581334b2ff))
* Unable to apply write authorization on relation property ([#941](https://github.com/plumier/plumier/issues/941)) ([39ff2b6](https://github.com/plumier/plumier/commit/39ff2b638d9cc5895b1368ef5e419e28c142b359))





## 1.0.1 (2021-05-18)

**Note:** Version bump only for package @plumier/core
