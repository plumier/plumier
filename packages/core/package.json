{
  "name": "@plumier/core",
  "version": "1.0.0-beta.9",
  "description": "Delightful Node.js Rest Framework",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "MVC",
    "Koa",
    "IoC",
    "Dependency Injection",
    "Web API",
    "JSON Service",
    "TDD"
  ],
  "scripts": {
    "compile": "tsc -p tsconfig.build.json"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@types/glob": "^7.1.1",
    "@types/koa": "^2.0.50",
    "chalk": "^2.4.2",
    "glob": "^7.1.4",
    "path-to-regexp": "^3.1.0",
    "tinspector": "^2.2.7",
    "tslib": "^1.10.0",
    "typedconverter": "^1.0.0"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "9d69b5621b7e2e777818e5e36c13ecba0c8d8204",
  "devDependencies": {
    "upath": "^1.2.0"
  }
}
