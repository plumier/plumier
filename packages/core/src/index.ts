import { val } from "typedconverter"
export { val }
export { AuthorizeCallback, AuthorizeMiddleware, RoleField, updateRouteAccess, Authorizer } from "./authorization";
export { HeaderPart, RequestPart, BindingDecorator, binder } from "./binder";
export { invoke } from "./middleware-pipeline";
export { response } from "./response";
export { analyzeRoutes, generateRoutes, printAnalysis } from "./route-generator";
export { router } from "./router";
export { ValidationMiddleware } from "./validator";
export { Class, consoleLog, findFilesRecursive, getChildValue, hasKeyOf, isCustomClass } from "./common";
export { AuthDecoratorImpl, authorize } from "./decorator.authorize";
export { bind } from "./decorator.bind";
export { domain, middleware } from "./decorator";
export { route, RouteDecoratorImpl } from "./decorator.route";
export { HttpStatus } from "./http-status";
export {
    ActionResult, Application, AuthorizeMetadataInfo, Configuration, DefaultFacility, CustomValidator,
    DependencyResolver, Facility, FileParser, FileUploadInfo, HttpMethod, HttpStatusError, Invocation, KoaMiddleware,
    Middleware, MiddlewareFunction, MiddlewareDecorator, MiddlewareUtil, PlumierApplication, PlumierConfiguration, RedirectActionResult,
    RouteContext, RouteInfo, RouteAnalyzerFunction, RouteAnalyzerIssue, ValidatorDecorator, ValidatorFunction,
    ValidatorInfo, ValidationError, errorMessage, AsyncValidatorResult, DefaultDependencyResolver
} from "./types";