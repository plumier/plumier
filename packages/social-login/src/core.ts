import {
    ActionResult,
    bind,
    CustomMiddleware,
    DefaultFacility,
    HttpCookie,
    HttpStatusError,
    Invocation,
    invoke,
    PlumierApplication,
    response,
    RouteInfo,
    RouteMetadata,
    VirtualRoute,
    Class
} from "@plumier/core"
import Axios, { AxiosError } from "axios"
import crypto from "crypto"
import Csrf from "csrf"
import debug from "debug"
import { Context } from "koa"
import OAuth from "oauth-1.0a"
import qs from "querystring"
import { decorateMethod } from "tinspector"

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

export enum CookieName {
    csrfSecret = "plum-oauth:csrf-secret",
    provider = "plum-oauth:provider"
}

export type SocialProvider = "Facebook" | "GitHub" | "Google" | "GitLab" | "Twitter"
export type OAuthVersion = "1.0a" | "2.0"
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export interface EndPoint {
    endpoint: string
    params: {}
}

export interface OAuthOptions {
    provider: SocialProvider
    clientId: string
    clientSecret: string
    token: EndPoint
    requestToken?: EndPoint
    profile: EndPoint & { transformer: (value: any) => Optional<OAuthUser, "provider"> }
    login: EndPoint
    oAuthVersion: OAuthVersion
}

export interface OAuthUser {
    provider: SocialProvider
    id: string,
    name: string,
    firstName: string,
    lastName: string,
    profilePicture: string,
    email?: string,
    gender?: string,
    dateOfBirth?: string
}

export interface OAuthProviderOption {
    clientId?: string,
    clientSecret?: string,
    loginEndPoint?: string,
    profileParams?: {}
}

interface ParserResult {
    user: Optional<OAuthUser, "provider">,
    profile: any,
    token: any
}

interface GeneratorResult {
    redirect: string,
    cookies: OAuthCookies
}

export type OAuth1Request = (opt: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => Promise<any>

// --------------------------------------------------------------------- //
// ------------------------------- HELPER ------------------------------ //
// --------------------------------------------------------------------- //

const log = {
    debug: debug("@plumier/social-login:debug"),
    error: debug("@plumier/social-login:error")
}

export function splitName(name: string | undefined) {
    if (!name) return { firstName: "", lastName: "" }
    const idx = name.lastIndexOf(" ")
    if (idx === -1) return { firstName: name, lastName: "" }
    const lastName = name.substr(idx + 1)
    const firstName = name.substring(0, idx)
    return { firstName, lastName }
}

class OAuthCookies {
    static async oAuth2(provider: SocialProvider) {
        const csrf = new Csrf()
        const secret = await csrf.secret()
        return new OAuthCookies(secret, provider)
    }

    constructor(public secret: string, private provider: SocialProvider) { }

    toHttpCookies(): HttpCookie[] {
        return [
            { key: CookieName.csrfSecret, value: this.secret },
            { key: CookieName.provider, value: this.provider },
        ]
    }
}

export function oAuth1ARequest(apiKey: string, apiKeySecret: string): OAuth1Request {
    var oauth = new OAuth({
        consumer: {
            key: apiKey,
            secret: apiKeySecret
        },
        signature_method: 'HMAC-SHA1',
        hash_function: function (base_string, key) {
            return crypto.createHmac('sha1', key).update(base_string).digest('base64');
        }
    });
    return async ({ url, method, data }: { url: string, method: "GET" | "POST", data: any }, token?: OAuth.Token) => {
        const headers = oauth.toHeader(oauth.authorize({ url, method, data }, token))
        const resp = await Axios({ url, method, headers, data: method === "GET" ? undefined : data })
        return resp.data
    }
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

export interface OAuthRedirectUriDecorator { name: "OAuthRedirectUriDecorator", provider: SocialProvider }

export function redirectUri(provider: SocialProvider | "General" = "General") {
    return decorateMethod(<OAuthRedirectUriDecorator>{ name: "OAuthRedirectUriDecorator", provider })
}

declare module "@plumier/core" {
    namespace bind {
        export function oAuthUser(): (target: any, name: string, index: number) => void
        export function oAuthToken(): (target: any, name: string, index: number) => void
        export function oAuthProfile(): (target: any, name: string, index: number) => void
    }
}

bind.oAuthUser = () => bind.custom(x => x.state.oAuthUser)
bind.oAuthToken = () => bind.custom(x => x.state.oAuthToken)
bind.oAuthProfile = () => bind.custom(x => x.state.oAuthProfile)



// --------------------------------------------------------------------- //
// ------------------------------ RESPONSE ----------------------------- //
// --------------------------------------------------------------------- // 

declare module "@plumier/core" {
    namespace response {
        function postMessage(message: any, origin?: string): ActionResult;
    }
}

response.postMessage = (message: any, origin?: string) => {
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                var message = '${JSON.stringify(message)}';
                var origin ${!!origin ? `= '${origin}'` : ""};
                (function(){
                    window.onbeforeunload = function () {
                        window.opener.postMessage({ status: "Close" }, origin || window.location.origin);
                    };
                    window.opener.postMessage(JSON.parse(message), origin || window.location.origin);
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

function clientRedirect(url: string) {
    // redirect client using JS to possibly set cookie before redirection
    return new ActionResult(`
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class="container"></div>
            <script type="text/javascript">
                const REDIRECT = '${url}';
                (function(){
                    window.location.href = REDIRECT;
                })()
            </script>
        </body>
    </html>
    `).setHeader("Content-Type", "text/html")
}

// --------------------------------------------------------------------- //
// ------------------------ LOGIN URL GENERATOR ------------------------ //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20LoginGenerator(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<GeneratorResult> {
    const params: any = { ...option.login.params, ...ctx.query }
    params.redirect_uri = ctx.origin + redirect_uri
    params.client_id = option.clientId
    const cookies = await OAuthCookies.oAuth2(option.provider)
    const csrf = new Csrf()
    params.state = csrf.create(cookies.secret)
    const redirect = option.login.endpoint + "?" + qs.stringify(params)
    return { redirect, cookies }
}

// oauth 1.0a
async function oAuth10aLoginGenerator(ctx: Context, opt: OAuthOptions, redirectUri: string): Promise<GeneratorResult> {
    const request = oAuth1ARequest(opt.clientId, opt.clientSecret)
    const rawTokens = await request({
        url: opt.requestToken!.endpoint,
        method: "POST",
        data: { oauth_callback: ctx.origin + redirectUri }
    })
    const reqTokens: any = qs.parse(rawTokens)
    const params = { oauth_token: reqTokens.oauth_token, ...opt.login.params, ...ctx.query }
    const redirect = `${opt.login.endpoint}?${qs.stringify(params)}`
    const cookies = new OAuthCookies(reqTokens.oauth_token, opt.provider)
    return { redirect, cookies }
}


// --------------------------------------------------------------------- //
// --------------------------- PROFILE PARSER -------------------------- //
// --------------------------------------------------------------------- //

// oauth 2.0
async function oAuth20ProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (!new Csrf().verify(secret, req.query.state)) {
        log.error("*** Invalid csrf token ***")
        log.error("CSRF Secret: %s", secret)
        log.error("CSRF Token: %s", req.query.state)
        throw new HttpStatusError(400)
    }
    const { clientId: client_id, clientSecret: client_secret } = option
    // request access token
    const { data: token } = await Axios.post<{ access_token: string }>(option.token.endpoint,
        { client_id, redirect_uri, client_secret, code: req.query.code, ...option.token.params },
        { headers: { Accept: 'application/json' } })
    // request profile
    const { data: profile } = await Axios.get(option.profile.endpoint, {
        params: option.profile.params,
        headers: { Authorization: `Bearer ${token.access_token}` }
    })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// oauth 1.0
async function oAuth10aProfileParser(ctx: Context, option: OAuthOptions, redirect_uri: string): Promise<ParserResult> {
    const req = ctx.request
    const secret = ctx.cookies.get(CookieName.csrfSecret)!
    if (secret !== req.query.oauth_token) {
        log.error("*** Invalid oauth_token ***")
        log.error("Saved token: %s", secret)
        log.error("Request token: %s", req.query.oauth_token)
        throw new HttpStatusError(400)
    }
    const request = oAuth1ARequest(option.clientId, option.clientSecret)
    const rawTokens = await request({
        url: option.token.endpoint, method: "POST",
        data: {
            oauth_token: req.query.oauth_token,
            oauth_verifier: req.query.oauth_verifier
        }
    })
    const token: any = qs.parse(rawTokens)
    const profile = await request({
        url: option.profile.endpoint, method: "GET",
        data: {}
    }, { key: token.oauth_token, secret: token.oauth_token_secret })
    const user = option.profile.transformer(profile)
    return { user, profile, token }
}

// --------------------------------------------------------------------- //
// ---------------------------- MIDDLEWARES ---------------------------- //
// --------------------------------------------------------------------- //

class OAuthLoginEndPointMiddleware implements CustomMiddleware {
    readonly generator = {
        "1.0a": oAuth10aLoginGenerator,
        "2.0": oAuth20LoginGenerator
    }

    constructor(private option: OAuthOptions, private loginPath: string, private redirectUri: string) { }

    async execute(invocation: Readonly<Invocation>): Promise<ActionResult> {
        if (invocation.ctx.path.toLowerCase() !== this.loginPath.toLowerCase()) return invocation.proceed()
        const { redirect, cookies } = await this.generator[this.option.oAuthVersion](invocation.ctx,
            this.option, this.redirectUri)
        return clientRedirect(redirect)
            .setCookie(cookies.toHttpCookies())
    }
}

class OAuthRedirectUriMiddleware implements CustomMiddleware {
    readonly parser = {
        "1.0a": oAuth10aProfileParser,
        "2.0": oAuth20ProfileParser
    }

    constructor(private option: OAuthOptions, private redirectUri: string) { }

    async execute(inv: Readonly<Invocation>): Promise<ActionResult> {
        const req = inv.ctx.request;
        if (inv.ctx.state.caller === "invoke") return inv.proceed()
        if (inv.ctx.path.toLocaleLowerCase() !== this.redirectUri.toLowerCase()) return inv.proceed()
        const provider = inv.ctx.cookies.get(CookieName.provider)
        if (provider !== this.option.provider) return inv.proceed()
        log.debug("Provider: %s", this.option.provider)
        const redirectUri = req.origin + req.path
        try {
            const result = await this.parser[this.option.oAuthVersion](inv.ctx, this.option, redirectUri)
            log.debug("OAuth User: %O", result.user)
            const provider = this.option.provider
            inv.ctx.state.oAuthUser = { ...result.user, provider }
            inv.ctx.state.oAuthToken = { ...result.token, provider }
            inv.ctx.state.oAuthProfile = { ...result.profile, provider }
        }
        catch (e) {
            log.error("*** OAuth Error ***", this.option.provider)
            log.error("Auth Code: %s", req.query.code)
            log.error("Redirect Uri: %s", redirectUri)
            if (e.response) {
                const error = (e as AxiosError)
                const response = error.response!;
                log.error("Request URL: %s %s", error.config.method, error.config.url)
                log.error("Request Data: %o", error.config.data)
                log.error("Request Headers: %o", error.config.headers)
                log.error("Error: %o", { status: response.status, message: response.statusText, data: response.data })
                throw new HttpStatusError(response.status, response.statusText)
            }
            throw e
        }
        return invoke(inv.ctx, inv.ctx.route!)
    }
}

// --------------------------------------------------------------------- //
// ------------------------------ FACILITY ----------------------------- //
// --------------------------------------------------------------------- //

export class OAuthProviderBaseFacility extends DefaultFacility {
    private option: Optional<OAuthOptions, "clientId" | "clientSecret">
    private loginEndpoint: string

    constructor(option: Optional<OAuthOptions, "clientId" | "clientSecret">, opt?: OAuthProviderOption) {
        super()
        this.option = {
            ...option,
            clientId: opt?.clientId,
            clientSecret: opt?.clientSecret,
        }
        if (opt && opt.profileParams)
            this.option.profile.params = { ...this.option.profile.params, ...opt.profileParams }
        this.loginEndpoint = opt?.loginEndPoint ?? `/auth/${this.option.provider.toLowerCase()}/login`
    }

    async generateRoutes(): Promise<VirtualRoute[]> {
        return [{
            kind: "VirtualRoute",
            method: "get",
            provider: this.constructor as Class,
            url: this.loginEndpoint,
            access: "Public",
            overridable: false,
            openApiOperation: {
                description: `Redirect request to ${this.option.provider} login page. This endpoint writes CSRF token on the client required to generate ${this.option.provider} login endpoint`,
                tags: ["Social Login"],
                parameters: [
                    {}
                ],
            }
        }]
    }

    getRedirectUri(routes: RouteMetadata[], provider: string) {
        return routes.find((x): x is RouteInfo => x.kind === "ActionRoute" && x.action.decorators
            .some((y: OAuthRedirectUriDecorator) => y.name === "OAuthRedirectUriDecorator" && y.provider === provider))
    }

    async initialize(app: Readonly<PlumierApplication>, routes: RouteMetadata[]) {
        const redirectUriRoute = this.getRedirectUri(routes, this.option.provider) ?? this.getRedirectUri(routes, "General")
        if (!redirectUriRoute) throw new Error(`OAuth: No ${this.option.provider} redirect uri handler found`)
        if (redirectUriRoute.url.search(":") > -1) throw new Error(`OAuth: Parameterized route is not supported on ${this.option.provider} callback uri`)
        if (redirectUriRoute.method !== "get") throw new Error(`OAuth: Redirect uri must have GET http method on ${redirectUriRoute.controller.name}.${redirectUriRoute.action.name}`)
        const clientIdKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_ID`
        const clientSecretKey = `PLUM_${this.option.provider.toUpperCase()}_CLIENT_SECRET`
        const clientId = this.option.clientId ?? process.env[clientIdKey]
        if (!clientId) throw new Error(`OAuth: Client id for ${this.option.provider} not provided`)
        const clientSecret = this.option.clientSecret ?? process.env[clientSecretKey]
        if (!clientSecret) throw new Error(`OAuth: Client secret for ${this.option.provider} not provided`)
        const option = { ...this.option, clientId, clientSecret }
        app.use(new OAuthLoginEndPointMiddleware(option, this.loginEndpoint, redirectUriRoute.url))
        app.use(new OAuthRedirectUriMiddleware(option, redirectUriRoute.url))
    }
}