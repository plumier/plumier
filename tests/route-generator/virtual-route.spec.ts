
import Plumier, { route, CustomMiddleware, Invocation, ActionResult } from "plumier"
import { Context } from "koa"
import { consoleLog, DefaultFacility, PlumierApplication, DefaultDependencyResolver } from '@plumier/core'

describe("Virtual Route", () => {

    it("Should able to print virtual route", async () => {
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .use(new MyMiddleware())
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should able to assigned multiple virtual route", async () => {
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        @route.virtual({ method: "get", url: "/route/second-route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .use(new MyMiddleware())
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should able print route that middleware registered by Facility on setup", async () => {
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        class MyFacility extends DefaultFacility {
            setup(app: Readonly<PlumierApplication>) {
                app.use(new MyMiddleware())
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .set(new MyFacility())
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should able print route that middleware registered by Facility on initialize", async () => {
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        class MyFacility extends DefaultFacility {
            async initialize(app: Readonly<PlumierApplication>) {
                app.use(new MyMiddleware())
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .set(new MyFacility())
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should able print route registered using string (DI)", async () => {
        const di = new DefaultDependencyResolver()
        const name = "myMiddleware"
        @di.register(name)
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .set({ dependencyResolver: di })
            .use(name)
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should able print route registered using symbol (DI)", async () => {
        const di = new DefaultDependencyResolver()
        const name = Symbol("myMiddleware")
        @di.register(name)
        @route.virtual({ method: "get", url: "/route/route", access: "Public" })
        class MyMiddleware implements CustomMiddleware {
            execute(invocation: Readonly<Invocation<Context>>): Promise<ActionResult> {
                return invocation.proceed()
            }
        }
        const mock = consoleLog.startMock()
        await new Plumier()
            .set({ dependencyResolver: di })
            .use(name)
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })


    it("Should not included middleware function", async () => {
        const mock = consoleLog.startMock()
        await new Plumier()
            .use(x => x.proceed())
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })

    it("Should not print anything if no virtual route found", async () => {
        const mock = consoleLog.startMock()
        await new Plumier()
            .initialize()
        expect(mock.mock.calls).toMatchSnapshot()
        consoleLog.clearMock()
    })
})