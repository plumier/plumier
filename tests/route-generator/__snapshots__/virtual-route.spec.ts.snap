// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virtual Route Should able print route registered using string (DI) 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should able print route registered using symbol (DI) 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should able print route that middleware registered by Facility on initialize 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should able print route that middleware registered by Facility on setup 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should able to assigned multiple virtual route 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/second-route",
  ],
  Array [
    "2. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should able to print virtual route 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
  Array [
    "Virtual Routes",
  ],
  Array [
    "These route might changed dynamically at runtime",
  ],
  Array [
    "1. MyMiddleware -> Public GET /route/route",
  ],
]
`;

exports[`Virtual Route Should not included middleware function 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
]
`;

exports[`Virtual Route Should not print anything if no virtual route found 1`] = `
Array [
  Array [],
  Array [
    "Route Analysis Report",
  ],
  Array [
    "1. AnimalController.get()   -> GET /animal/get",
  ],
  Array [
    "2. BeastController.get()    -> GET /beast/get",
  ],
  Array [
    "3. CreatureController.get() -> GET /creature/get",
  ],
  Array [],
]
`;
